### ISSUE TITLE
Test docgen code at a granular level

### ISSUE THREAD

**Issue by magicDGS** [broadinstitute/barclay#71]
From @cmnbroad (https://github.com/broadinstitute/barclay/pull/68#issuecomment-306683475):
> One high-level request is that we find a way to incrementally unit-test new functionality like this at a granular level. We currently only have coarse-grained, file-based integration tests, and I'd like to find a way to avoid proliferating a new set of test files with each new feature, in addition to changing many/all of the existing ones. @magicDGS any thoughts on how we can address that ?

**Comment by magicDGS** [comment]
Regarding this issue, I can tell you what it is the idea that I'm trying to implement in my own code:

1. As in the current test implementation, I execute the javadoc to check that the whole doclet code is working and the templates are correct. In Barclay, I suggest to keep just the test for the default doclet/templates with some basic annotated classes, but remove the custom `TestDoclet` and associated templates.
2. For concrete features that may or may not be included in the basic doclet generation test, I'm trying to use a mocking framework (e.g. [Mockito](http://site.mockito.org/)) for `ClassDoc`. Like that, I can create `DocWorkUnit`s without the need of executing the javadoc, and testing methods for docgen classes in a granular way.

Is this reasonable for you, @cmnbroad?

**Comment by cmnbroad** [comment]
Yes, a mocking framework is probably the right way to do this. Do you already have code that does this ? I'd like to solve this problem in Barclay before we add any more new features.

**Comment by magicDGS** [comment]
I was exploring this a bit in my toolkit and I implemented some minor tests as a proof-of-concept. If you like, I can add a couple of tests with Mockito and remove the `TestDoclet` code by the end of next week. Let me know if that fits you!

**Comment by cmnbroad** [comment]
@magicDGS That would be very useful. I wouldn't remove the TestDoclet test though, since we definitely need to test both default and custom doclet cases. At some point we may have enough mocked unit tests that we can remove one of them, but for now I think we should keep both. At least this will give us a start in the direction of less file bloat.
